File: leetcode_problems/23. 合并K个升序链表.go
package main

func main() {}

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeKLists(lists []*ListNode) *ListNode {
	length := len(lists)
	if length < 1 {
		return nil
	}
	if length == 1 {
		return lists[0]
	}
	num := length / 2
	left := mergeKLists(lists[:num])
	right := mergeKLists(lists[num:])
	return subMerge(left, right)
}

func subMerge(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil {
		return l2
	}
	if l2 == nil {
		return l1
	}
	if l1.Val < l2.Val {
		l1.Next = subMerge(l1.Next, l2)
		return l1
	}
	l2.Next = subMerge(l1, l2.Next)
	return l2
}

File: leetcode_problems/23. 合并K个升序链表.ts
class ListNode {
	val: number
	next: ListNode | null
	constructor(val?: number, next?: ListNode | null) {
		this.val = val === undefined ? 0 : val
		this.next = next === undefined ? null : next
	}
}
function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
	const len = lists.length
	if (len < 1) {
		return null
	}
	if (len === 1) {
		return lists[0]
	}
	let mid = Math.floor(len / 2)
	let head: Array<List> = [],
		tail: Array<List> = []
	lists.map((v, index) => {
		if (index < mid) {
			//若等于则会死循环，因为始终2个分不下去。
			head.push(v)
		} else {
			tail.push(v)
		}
	})
	let left = mergeKLists(head)
	let right = mergeKLists(tail)
	return subMerge(left, right)
}

type List = ListNode | null
const subMerge = (l1: List, l2: List): List => {
	if (!l1) {
		return l2
	}
	if (!l2) {
		return l1
	}
	if (l1.val < l2.val) {
		l1.next = subMerge(l1.next, l2)
		return l1
	} else {
		l2.next = subMerge(l2.next, l1)
		return l2
	}
}
